File: weisi/src/main/java/com/cjq/lib/weisi/communicator/tcp/TcpSocket.java
package com.cjq.lib.weisi.communicator.tcp;

import android.support.annotation.NonNull;

import com.cjq.lib.weisi.communicator.Communicator;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;

/**
 * Created by CJQ on 2018/3/9.
 */

public class TcpSocket implements Communicator {

    private Socket mSocket;
    private InputStream mReader;
    private OutputStream mWriter;

    public TcpSocket(@NonNull Socket socket) throws IOException {
        mSocket = socket;
        mReader = mSocket.getInputStream();
        mWriter = mSocket.getOutputStream();
    }

    public void write(byte[] src) throws IOException {
        write(src, 0, src.length);
    }

    public void write(byte[] src, int offset, int length) throws IOException {
        mWriter.write(src, offset, length);
    }

    public int receive(byte[] dst, int offset, int length) throws IOException {
        return mReader.read(dst, offset, length);
    }

    @Override
    public int read(byte[] dst, int offset, int length) throws IOException {
        int expectLen = mReader.available();
        return expectLen > 0
                ? receive(dst, offset, length)
                : expectLen;
    }

    @Override
    public boolean canRead() {
        return mSocket.isConnected();
    }

    @Override
    public void stopRead() throws IOException {
    }

    public void close() throws IOException {
        if (mSocket != null && !mSocket.isClosed()) {
            mReader.close();
            mWriter.close();
            mSocket.close();
        }
    }
}

File: wsb/src/main/java/com/wsn/lib/wsb/communicator/tcp/TcpSocket.kt
package com.wsn.lib.wsb.communicator.tcp


import com.wsn.lib.wsb.communicator.Communicator

import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.Socket

/**
 * Created by CJQ on 2018/3/9.
 */

class TcpSocket @Throws(IOException::class)
constructor(private val socket: Socket) : Communicator {
    private val reader = socket.getInputStream()
    private val writer = socket.getOutputStream()

    fun isConnected() = socket.isConnected

    @Throws(IOException::class)
    @JvmOverloads
    fun write(src: ByteArray, offset: Int = 0, length: Int = src.size) {
        writer.write(src, offset, length)
    }

    @Throws(IOException::class)
    fun receive(dst: ByteArray, offset: Int, length: Int): Int {
        return reader.read(dst, offset, length)
    }

    @Throws(IOException::class)
    override fun read(dst: ByteArray, offset: Int, length: Int): Int {
        val expectLen = reader.available()
        return if (expectLen > 0)
            receive(dst, offset, length)
        else
            expectLen
    }

    override fun canRead(): Boolean {
        return isConnected()
    }

    @Throws(IOException::class)
    override fun stopRead() {
    }

    @Throws(IOException::class)
    fun close() {
        if (!socket.isClosed) {
            reader.close()
            writer.close()
            socket.close()
        }
    }
}
